src/Network/MQTT/Client.hs

```diff
@@ -129,6 +133,17 @@ connectURI cfg uri = do
     up x = let (u,r) = break (== ':') (init x) in
              (Just (unEscapeString u), if r == "" then Nothing else Just (unEscapeString $ tail r))
 
+-- | Memory 'E.bracket' around 'connectURI' and 'disconnect'
+withConnectURI :: MQTTConfig -> URI -> (MQTTClient -> IO c) -> IO c
+withConnectURI cfg uri = E.bracket (connectURI cfg uri) disconnect
```

Comment:
My concern about bracketing these is it implies disconnect is a good idea.  This will cause the lastwill to not be sent in a client program crash.  I'd think it'd be better to leave that evidence in the case of failure.

Reply:
I think perhaps something along the lines of:

```
withClientURI :: MQTTConfig -> URI -> (MQTTClient -> IO c) -> IO c
withClientURI cfg uri f = E.bracketOnError
                          (connectURI cfg uri)
                          (\c -> disconnect c DiscoUnspecifiedError mempty)
                          (\c -> f c >>= \x -> normalDisconnect c >> pure x)
```

would be good.  The idea is that we want to communicate an unspecified error when there's an error (we could actually put an error string in the properties if we needed to), but send a normal disconnect in the normal completion case.  This should make the will code work as expected.

Reply:
This doesn't cause the last-will to be sent by the broker. The only way the last-will is sent when sending a `DISCONNECT` packet instead of ending the connection abruptly is by disconnection with the `DiscoDisconnectWithWill` reason but this requires MQTT v5.

Reply:
Makes sense.  I think we could work around the protocol version, but I'd want my will by default.  I think *not* including a will would be some kind of state transform on the client where I intend a clean shutdown.